Documentation/ch-scrub-intro.rst

```diff
@@ -6,14 +6,27 @@ On filesystems that use replicated block group profiles (e.g. raid1), scrub will
 also automatically repair any damage by default by copying verified good data
 from one of the other replicas.
 
+.. warning::
+   Files with the NOCOW (``chattr +C``) attribute imply ``nodatasum``.
+   Errors in these files cannot be detected or corrected by scrub. This means
+   the NOCOW attribute is not currently safe in the presence of replicas since
```

Comment:
I'd say the `not safe` claims is not ideal.
This means you're calling all file systems without data checksum unsafe.

And even for btrfs, there are known IO patterns (abuse) that can lead to mismatch data checksum directly.
Namely incorrectly handled writeback (modify the direct IO memory while it's still under writeback).

Reply:
Sure, let me make sure there’s a mutual understanding of both the purpose of this warning and the operation of btrfs so I can update the wording to make sure it is communicating the right thing without being too inaccurate.

The purpose of this warning is to prevent users from experiencing data loss when they think btrfs is protecting their files against single replica corruption. It is not clear right now that simply having a healthy replica is insufficient to avoid data loss when using NOCOW. For example, [this Reddit thread](https://old.reddit.com/r/btrfs/comments/xki69r/nodatacow_and_data_loss/).

In my case, I had one journald file end up corrupted, and did not understand how that was possible unless my *whole* filesystem was corrupted, until someone else pointed out that journald sets those files to NOCOW, and that is why *only* that file ended up being irreparably damaged.

With NOCOW, my understanding is that btrfs does not even meet standards of traditional raid1 systems which would have marked the unhealthy replica as tainted and only read from the healthy one, thus avoiding corruption in this scenario. As such, I pick this language to express that at least with single profile, your file might be silently corrupted if there is a crash, but you know that already because you have no replica. With replicas, it is “not safe” because having a copy committed to a healthy replica does *not* (apparently, unless it’s a bug) prevent corruption.

It is not clear to me how btrfs handles NOCOW exactly since e.g. #134 talks about multiple copies being possible to restore, but I would expect that only to be true until btrfs notices the metadata is inconsistent and fixes it up using the one with the correct checksum/transid and marks the blocks from the bad metadata as free again. (Clearly I do not understand the architecture of btrfs very much at all and am trying to potentially apply knowledge from other domains that do not apply, so please correct and explain so I can understand and be smarter :-))

Reply:
> With NOCOW, my understanding is that btrfs does not even meet standards of traditional raid1 systems which would have marked the unhealthy replica as tainted and only read from the healthy one, thus avoiding corruption in this scenario.

For NOCOW, the other RAID1 implementations go marking the bad device with extra flags, and I believe the btrfs part can indeed be improved.

But if the data corruption happens without really triggering any read error (e.g. for whatever reason the data on-disk just changed, without triggering a read error), then it's no better than NODATACOW btrfs RAID1, since just like NODATACOW btrfs, the block layer RAID1 won't detect any error, thus go the same path as btrfs (to go load balance the read)


That's the main reason I won't call it unsafe, because for above case the block layer RAID1 implementation is just as bad as btrfs.

However I also won't decline that btrfs is pretty bad at detecting unreliable disks, thus it won't avoid reading that bad disk.

I'd prefer to make it clear that, btrfs doesn't have any ability to detect bad disks (which can be very complex by itself, and I'm very curious how other RAID1 handles it), so for that NODATACOW + bad disks case, users will get data corruption, that's correct.

But it doesn't mean the other RAID1 is that safe, they just mask their lack of data csum by detecting if the disk is bad.

> It is not clear to me how btrfs handles NOCOW exactly since e.g. https://github.com/kdave/btrfs-progs/issues/134 talks about multiple copies being possible to restore, but I would expect that only to be true until btrfs notices the metadata is inconsistent and fixes it up using the one with the correct checksum/transid and marks the blocks from the bad metadata as free again.

First of all, NOCOW can only be applied to data, thus we always call it NODATACOW.
Metadata is *always* checksumed, and *always* COWed. So metadata is always safe.
(as long as the disks implement the barrier operations correctly)

And unfortunately even with RAID1C3/C4, btrfs won't compare all the mirrors and choose the correct one (the one matches the most mirrors), mostly because it will bring a huge performance drop.

It's back to the bad disks detection which is lacking in btrfs, unfortunately.

Reply:
> For NOCOW, the other RAID1 implementations go marking the bad device with extra flags, and I believe the btrfs part can indeed be improved.

Yes, if some enhancement like this is done then I think btrfs raid1 will be guaranteed to never be worse than traditional software raid1 in all failure modes, which would be great.

> But if the data corruption happens without really triggering any read error (e.g. for whatever reason the data on-disk just changed, without triggering a read error), then it's no better than NODATACOW btrfs RAID1

Sure, this is why I personally opted to use btrfs raid1 instead of mdadm. :-) If I understand correctly, it is only the combination of NOCOW + media failure where btrfs ends up in a worse state due to what you mention about bad disk detection and then load balancing reads the wrong data. Unfortunately media failure is probably the most common failure mode. :-(

> just like NODATACOW btrfs, the block layer RAID1 won't detect any error, thus go the same path as btrfs (to go load balance the read)

So I think this is right, but just to confirm: if a user purges OS file cache and retries, eventually they will read the good data, since the metadata is kept consistent on both mirrors always, and the corruption is just due to how one disk did not write out the data blocks pointed by the metadata? This makes a lot of sense and is straightforward enough that I may provide this additional context in the documentation.

If I got that right, is it possible to force a read from one device (without mounting degraded with only the single healthy device connected)? It would be good to mention in the documentation if this is the case. It would be very sad if users are “losing” data in this situation only because they can’t tell btrfs to force read/overwrite the file from the known good disk. I understand #482 is a request to add something akin to this to scrub but wonder if it is possible to do it in an ugly way now?

I did not remove the reference to the bad journal file on my filesystem yet, so I wonder if I could test this. So far it seems like either the file got is cached, or I get unlucky and never hit the right PID to balance to the healthy device, or journald read the bad log on reboot, processed it into garbage, then deleted the original inode, so now it is actually a file that is just bad forever. I have no idea!

> I'd prefer to make it clear that, btrfs doesn't have any ability to detect bad disks (which can be very complex by itself, and I'm very curious how other RAID1 handles it), so for that NODATACOW + bad disks case, users will get data corruption, that's correct.

OK, I think I can rewrite to be more like this since that feels better to you and it seems OK to me.

> But it doesn't mean the other RAID1 is that safe, they just mask their lack of data csum by detecting if the disk is bad.

Yes, I understand. The “not safe” phrase was meant only by comparison to btrfs with CoW data mode, not to any others. :-)

> First of all, NOCOW can only be applied to data, thus we always call it NODATACOW.

Sure, this makes sense. I used the terminology NOCOW here because that is how I see the `+C` attribute described elsewhere (or else in the `chattr` man page, `No_COW`). I think it will be good to also clarify in the documentation so will try to work this in too.

> And unfortunately even with RAID1C3/C4, btrfs won't compare all the mirrors and choose the correct one (the one matches the most mirrors), mostly because it will bring a huge performance drop.

This also seems like a good information to write in the docs, just for anyone who wonders about it, so I will try to work that in too.

Changes should come tomorrow and I will drop a note when that happens.

Reply:
> if a user purges OS file cache and retries, eventually they will read the good data, since the metadata is kept consistent on all replicas always, and the corruption is just due to how one disk did not write out the data blocks pointed by the metadata? 

Firstly for metadata read, if we hit any bad (csum/generation/level mismatch), we immediately try the other mirrors. Either we got a good one, then writeback the good content to the bad mirror. Or we exhaust all copies and return error.

Then for the data, if it has checksum (and checksum is stored in metadata), we go the same way above for repair.
If it has no checksum, we just return the content we read out (which mirror is based on the pid based load balance).

So for data with checksum, we will always grab a good mirror if possible, and a read on the bad mirror will also cure the bad mirror (if there is a good one).

> If I got that right, is it possible to force a read from one device (without mounting degraded with only the single healthy device connected)?
Following my previous answer, if there is data checksum, no force needed, we immediately repair any bad data copy if we have any good mirror remaining.
(For metadata it's always repaired since it has csum/level/generation checks)

For data without csum, if you can get a good mirror really depends on the pid of the read process.
It's not as good as "forcing btrfs to read certain copy", but if you really want to go that path, it's still possible using "btrfs-map-logical" command to manually read certain mirror of a logical bytenr, and fetch the good copy.

But NODATACOW also means, we won't COW the data write, meaning if your process read out the bad content, modify some bytes, and write it back, the good copy is also gone now.

So for your journal case, I do not think it's a good chance to get the good mirror back, because later writes are definitely going to screw the good mirror.

Reply:
Currently btrfs doesn't have any way to reliably resync data between mirrors if a device goes offline and comes back while mounted.  Excluding recovery methods which require umounting the filesystem, the only tool we have for this case is scrub, and scrub's reliability for data blocks is limited to the csum collision rate (this isn't a problem for metadata blocks because there are other header fields to indicate a lost write, so a csum collision is not sufficient to create an undetected error in metadata).  For crc32c csums, that's an error rate of 1 in 16 TiB, high enough for multiple undetected errors per drive in the worst case on modern hardware.  For nodatacow files, the undetected error rate on the offline drive is always 100% of errors, because scrub doesn't do anything with nodatacow files at all.

In mdadm it's possible to explicitly mark a device as bad and then force a resync in place (i.e. replacing the drive with itself).  In this case, mdadm reads only from the good devices, so it restores correct data reliably to the bad ones (assuming there's only one corrupt drive and the user knows which one that is, but this is quite common when a single controller or cable fails).

It's possible to do resync in place in btrfs by patching `btrfs replace` to skip the mounted check, then running `btrfs replace start -r` through a loop device to avoid the kernel's same-device check.  We do this whenever we need to get one drive reliably resilvered after an online corruption event.  It would be nice if upstream `btrfs replace` supported this operation directly.

Currently the kernel code explicitly forbids a replace target from being the same device as the source,  and btrfs-progs userspace checks to see if the device is an existing device in the filesystem before allowing replace, without making an exception for self-replace.  Both of those are bugs that should be fixed, or scrub should be extended to support resilvering explicitly.

Reply:
> For crc32c csums, that's an error rate of 1 in 16 TiB, high enough for multiple undetected errors per drive in the worst case on modern hardware.

I am curious about the worst case here. Would it be that all data blocks of a large filesystem are rewritten between device disappearing and reappearing? I would naïvely assume that the number of changed data blocks would typically be low enough that the chance of a collision on any actually changed block due to a transient dropout would be sufficiently low to not have to worry in practice, but I don’t really trust my judgment at all about anything.

Notwithstanding, the ability to *guarantee* a zero error rate by resilvering back certainly sounds preferable to me when there’s no need to take that risk, if that makes sense, and is what you are saying.

I don’t like the feeling I have had during this hardware event that btrfs volume management is optimised for making changes to healthy filesystems and unexpectedly difficult to reason about in a failure scenario. I’m not sure where is the best place to discuss this so maybe something can be improved.

Reply:
I’ve added two new commits to this patch set, one that hopefully satisfies this feedback, and then a new warning in the section that talks about using balance to remove a device from a redundant set. As always I am not super confident that I am speaking accurately, so please check and let me know about any inaccuracy. Thanks!

Reply:
> Would it be that all data blocks of a large filesystem are rewritten between device disappearing and reappearing?

Big corruption events do happen, but they are typically caused by failures between btrfs and the devices, not the devices themselves.  A couple of examples from earlier this year:

1.  An unmonitored filesystem (e.g. a home user with a NAS) might have a drive go offline that is not noticed until weeks later when it comes back online after a reboot.  Daily data rates these days can be large enough to get to double-digit TB in that time, and they're only going up.

2.  We recently had an incident where `pvmove` went badly wrong during a routine upgrade of an 18 TB LV device under a btrfs filesystem.  Everything seemed normal:  pvmove finished with no error, iostat reported a little over 18 TB read and 18 TB written during pvmove...but it turned out that only about 6 TB of data had actually been moved to the correct location.  This wasn't discovered until several hours into the post-pvmove scrub.  From the btrfs point of view, there were 12 TB of randomly corrupted data on the upgraded device.  We weren't going to roll the dice on a 75% chance of an undetected error with a scrub, so we aborted the scrub and forced a full device replace instead.

The risk for datasum files can be substantially reduced by using xxhash64 or larger instead of crc32c for csums, but this requires preparation in advance (best at mkfs time), as it's not convenient to change csum method on the fly.

> the ability to _guarantee_ a zero error rate by resilvering back certainly sounds preferable

For nodatacow files, it's not a matter of preference--resilvering or replacing the bad drive is the only recovery method that can work _at all._  (and it only works if the problem is confined to a single drive and the user knows which drive is affected)

> unexpectedly difficult to reason about in a failure scenario

It's easy to reason about failure scenarios if you're familiar with the implementation details, but that familiarity is also the _only_ way to reason about failure scenarios.  btrfs has unique architectural features, so knowledge of other filesystems or RAID schemes doesn't necessarily apply (e.g. in ext4 or xfs on mdadm or lvm, a user can't exclude an individual file from replication or dm-integrity, but on btrfs they can).  There are some things that can be improved, but also some things that are inevitable fallout of mutually exclusive goals.

Reply:
> It's easy to reason about failure scenarios if you're familiar with the implementation details, but that familiarity is also the only way to reason about failure scenarios.

Sure, some of it is the different design approach and that’s OK, especially since those differences mean that btrfs fixes many fundamental problems of other filesystems.

To me, the unexpected difficulty comes from how I think this particular user story shouldn’t require anyone to become a filesystem implementation expert: “I want my system to remain online and usable even if one of two disks fails, without any data corruption due to the failure. I want to be able to replace the failed disk using a single command (or two commands if partitioning cannot be automated), and have confidence that this one command will correctly restore everything to the intended level of redundancy. I do not want to feel confused about what steps are safe or not, nor inadvertently run commands that will non-obviously cause or increase the risk of data loss or downtime.”

In the case of mdadm, the failing disk would be flagged and write-intent bitmap would be used to mark what data didn’t make it to the failed disk. No corruption[*] or confusion about what steps to take. Just remove old disk, insert new disk, partition, run `mdadm --add /dev/X` or `mdadm --replace /dev/Y --with /dev/X`, and wait. You can even mark the failing disk to not be used at all in advance if it is causing problems. In any case, one hardly needs to know anything about the implementation, and the parts that need to be known are simple.

In the case of btrfs, it does not notice the disk failed at all, so starts using it again like normal after a reboot even though it is unhealthy. At least some `nodatasum` data gets lost probably immediately due to unlucky load balancing. Correction of remaining `datasum` data is apparently possible only by checksum correction, and now we talk about how probably that isn’t too safe either except for metadata since btrfs default is crc32c which has very low collision resistance?

The actual recovery requires either adding an extra disk first, or using `-o degraded`. If adding an extra disk first, when running `btrfs replace`, the default behaviour is not with `-r` flag, so if there is any bad `nodatasum` data on the failing disk or `datasum` data where there was a hash collision, by default, corrupted data will be written to the new replacement, instead of good data from the healthy device. (Does `-r` absolutely guarantee no reads from the unhealthy device in this case, or is it mandatory to pull the unhealthy device and use the `-o degraded` path?)

If using `-o degraded` on a root filesystem, it probably also needs to be done from a recovery environment, since setting `-o degraded` in advance sounds like one good way to accidentally get a split brain if there is ever any number of other types of fault that cause the healthy disk to temporarily not appear. There’s also conflicting information on the internet about whether this mount option is a risky one-chance operation, or if that was just a bug that got fixed (official docs don’t seem to say one way or the other). After getting to this point, the user then needs to know what devid to use since the old disk won’t exist to replace; according to [this wiki](https://wiki.tnonline.net/w/Btrfs/Replacing_a_disk) (which contains more important-seeming warnings that don‘t seem to be listed in the official documentation), the user has to run `btrfs device usage` to find the missing disk ID, then put that into `btrfs replace`.

In both cases, the user must remember afterward not to forget to run `btrfs balance`, or else now some of their data is a time bomb.

Especially as btrfs becomes more mainstream and takes over as the default filesystem for some distributions, I really feel like it needs to be about as simple, and at least as reliable, as mdadm in this failure scenario. I don’t see why there would not be a path to get there, but some of these problems seem to have been raised [in at least 2018](https://www.mail-archive.com/linux-btrfs@vger.kernel.org/msg78254.html), so it feels bad to encounter the same thing six years later.

OK, thanks for listening to my TED talk. I think everything I wrote is accurate but as usual please correct any misconceptions or, if you would, unanswered questions. My apologies if this comes across as hostile; again, I think btrfs is doing more right things than wrong things, and so when btrfs makes most things in daily operation quite easy, the unexpected difficulty of a disaster recovery is heightened. The double-edged sword of success :-)

[*] Pedantic hedging here: I don’t know this for sure, I have never personally experienced this situation, maybe it breaks too, but all evidence suggests it would do the right thing.

Reply:
>  chattr -C result is “Operation not permitted” on this file and three others, not sure if that’s intended or why it would be restricted, no output anywhere says why.

cow/nocow mode can not be toggled on files that have data. The man page for `chattr` states: 
> Note:  For btrfs, the ’C’ flag should be set on new or empty files.

Reply:
>  The man page for `chattr` states: […]

Sure, but I was trying to *clear* the flag, not set it, you see :-)

By the way, thank you for your wiki content, I hope that much of it makes its way into the official documentation someday; the examples for how to handle disk replacement were the best instructions I found and I think the official documentation deserves that level of clarity.

Reply:
> > The man page for `chattr` states: […]
> 
> Sure, but I was trying to _clear_ the flag, not set it, you see :-)

True, we could always improve documentation and be better at providing good UX for users.

> By the way, thank you for your wiki content, I hope that much of it makes its way into the official documentation someday; it is quite well organised and the examples for how to handle disk replacement were the best instructions I found and I think the official documentation deserves that level of clarity.

Thanks! I am glad the wiki is useful. I started it mostly as a way to learn Btrfs and write down how to do things in an easy way. It's great to see so much improvement that has gone into the official documentation since then 🙏

> With NOCOW, my understanding is that btrfs does not even meet standards of traditional raid1 systems which would have marked the unhealthy replica as tainted and only read from the healthy one.

IMHO, it may not always be optimal to mark a disk as bad when it clearly can read some data because it could happen that data on the good disk also becomes corrupt. Still, some logic to determine how and when a disk should be dropped/marked bad would be useful. Probably best implemented as a user-space daemon that monitors error rate, etc. 

Also, to clarify that md/dm raid also would not detect corruption. They would detect read-errors, but that would Btrfs too. 

> And unfortunately even with RAID1C3/C4, btrfs won't compare all the mirrors and choose the correct one (the one matches the most mirrors), mostly because it will bring a huge performance drop.

Such a feature could probably be added to scrub. It could let users detect and choose which copy to preserve. https://github.com/kdave/btrfs-progs/issues/482

Reply:
> IMHO, it may not always be optimal to mark a disk as bad when it clearly can read some data because it could happen that data on the good disk also becomes corrupt. Still, some logic to determine how and when a disk should be dropped/marked bad would be useful. Probably best implemented as a user-space daemon that monitors error rate, etc.

Sure, I can see arguments on both sides of this, but it can only really be decided once btrfs has a functioning write-intent bitmap. Reads from a failed device—which mdadm will do during recovery when the failed device is still accessible to avoid stressing other disks into failure—are only safe with a write-intent bitmap. It is unsound for btrfs to continue reading from a disk with failed write because it has no write-intent bitmap to skip/fix outdated blocks. This is absolutely true for nodatasum data, but it is also true even for datasum data due to possibility of hash collision, as mentioned by @Zygo—though the risk is low enough *in practice* that you can probably get away with it a lot of the time (I managed).

> Also, to clarify that md/dm raid also would not detect corruption. They would detect read-errors, but that would Btrfs too.

mdadm maintains write-intent bitmap, so does detect one form of “corruption” that btrfs does not, which is “corruption” due to writes that only made it to one device.

I tried to drive it home on my mailing list post, and I try to do the same here: btrfs needs write-intent bitmap urgently. Without write-intent bitmap, btrfs is like a super-advanced car with the most advanced airbags and collision-avoidance sensor technology, but no seatbelts.